x/session/sessionNodes.go

```diff
@@ -0,0 +1,141 @@
+	}
+	// using mock world state
+	allActiveNodes, err := fixtures.GetNodes()
+	if err != nil {
+		return nil, err
+	}
+	// session nodes are just a wrapper around node slice
+	var result SessionNodes
+	// node distance is just a node with a computational field attached to it
+	var xorResult []NodeDistance
+	// ensure there is atleast the minimum amount of nodes
+	if len(*allActiveNodes) < SESSIONNODECOUNT {
+		return nil, InsufficientNodesError
+	}
+	// filter `allActiveNodes` by the HASH(nonNativeChain)
+	result, err = filter(AllActiveNodes(*allActiveNodes), nonNativeChain)
+	if err != nil {
+		return nil, err
+	}
+	// xor each node's public key and session key
+	// return NodeDistance array to be ordered
+	xorResult, err = xor(result, sessionKey)
+	if err != nil {
+		return nil, err
+	}
+	// sort the nodes based off of distance
+	result = revSort(xorResult)
+
+	// return the top 5 nodes
+	return result[0:5], nil
+}
+
+// filter the nodes by non native chain
+func filter(allActiveNodes AllActiveNodes, nonNativeChainHash []byte) (SessionNodes, error) {
+	var result SessionNodes
+	for _, node := range allActiveNodes {
+		if !node.SupportedChains.Contains(hex.EncodeToString(nonNativeChainHash)) {
+			continue
+		}
+		result = append(result, node)
+	}
+	if len(result) == 0 {
+		return nil, InsufficientNodesError
+	}
+	return result, nil
+}
+
+// xor the sessionNodes.publicKey against the sessionKey to find the computationally
+// closest nodes
+func xor(sessionNodes SessionNodes, sessionkey SessionKey) (NodeDistances, error) {
+	var keyLength = len(sessionkey)
+	result := make([]NodeDistance, len(sessionNodes))
+	// for every node, find the distance between it's pubkey and the sesskey
+	for index, node := range sessionNodes {
+		pubKey, err := node.PubKey.Bytes()
+		pubKeyHash := crypto.Hash(pubKey) // currently hashing public key but could easily just take the first n bytes to compare
+		if err != nil {
+			return nil, err
+		}
+		if len(pubKeyHash) != keyLength {
+			return nil, MismatchedByteArraysError
+		}
+		result[index].Node = node
+		result[index].distance = make([]byte, keyLength)
+		for i := 0; i < keyLength; i++ {
+			result[index].distance[i] = pubKeyHash[i] ^ sessionkey[i]
+		}
+	}
+	return result, nil
+}
+
+// sort the nodes by shortest computational distance
+func revSort(sessionNodes NodeDistances) SessionNodes {
+	result := make(SessionNodes, len(sessionNodes))
+	sort.Sort(sort.Reverse(sessionNodes))
+	for _, node := range sessionNodes {
+		result = append(result, node.Node)
+	}
+	return result
+}
+
+// returns the length of the node pool -> needed for sort.Sort() interface
+func (n NodeDistances) Len() int { return len(n) }
+
+// swaps two elements in the node pool -> needed for sort.Sort() interface
+func (n NodeDistances) Swap(i, j int) { n[i], n[j] = n[j], n[i] }
+
+// returns if node i is less than node j by XOR value
+// it assumes big endian encoding
+func (n NodeDistances) Less(i, j int) bool {
```

Comment:
Where is this being used?

Reply:
It's part of the interface needed for sort.Sort

Reply:
Got it, a comment about that like you did in the 2 methods above would help identify that.